File: gek-core/src/main/java/xyz/fsgek/common/reflect/TypeRef.java
package xyz.fsgek.common.reflect;

import xyz.fsgek.common.base.Fs;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * Type reference usually to obtain type instance for java, for examples:
 * <p>
 * String.class:
 * <pre>
 *     TypeRef&lt;String&gt; typeRef = new TypeRef&lt;String&gt;(){};
 *     Type type = typeRef.getType();
 * </pre>
 * <p>
 * Generic type Map&lt;String, String&gt;:
 * <pre>
 *     TypeRef&lt;Map&lt;String, String&gt;&gt; typeRef = new TypeRef&lt;Map&lt;String, String&gt;&gt;(){};
 *     Type type = typeRef.getType();
 * </pre>
 *
 * @author fredsuvn
 */
public abstract class TypeRef<T> {

    /**
     * Actual runtime type.
     */
    private Type type;

    /**
     * Empty constructor, used to get a generic type.
     */
    protected TypeRef() {
    }

    private Type reflectToActualType() {
        Type generic = getClass().getGenericSuperclass();
        if (generic instanceof ParameterizedType) {
            ParameterizedType p = (ParameterizedType) generic;
            if (Fs.equals(p.getRawType(), TypeRef.class)) {
                return p.getActualTypeArguments()[0];
            }
        }
        ParameterizedType parameterizedType = FsReflect.getGenericSuperType(generic, TypeRef.class);
        if (parameterizedType == null) {
            throw new IllegalStateException("Current type is not subtype of TypeRef: " + getClass().getName());
        }
        return parameterizedType.getActualTypeArguments()[0];
    }

    /**
     * Returns type of this type-ref.
     *
     * @return type of this type-ref
     */
    public Type getType() {
        if (type == null) {
            type = reflectToActualType();
        }
        return type;
    }

    /**
     * Returns {@link #getType()} as {@link ParameterizedType} of this type-ref.
     *
     * @return type as {@link ParameterizedType} of this type-ref
     */
    public ParameterizedType asParameterized() {
        return (ParameterizedType) getType();
    }

    /**
     * Returns {@link #getType()} as {@link Class} of this type-ref.
     *
     * @return type as {@link Class} of this type-ref
     */
    public Class<T> asClass() {
        return (Class<T>) getType();
    }
}

File: boat-common/src/main/kotlin/xyz/srclab/common/reflect/TypeRef.kt
package xyz.srclab.common.reflect

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

abstract class TypeRef<T> {

    @get:JvmName("type")
    val type: Type

    protected constructor() {
        type = reflectGenericSuperclass()
    }

    protected constructor(type: Type) {
        this.type = type
    }

    private fun reflectGenericSuperclass(): Type {
        val generic = this.javaClass.findGenericSuperclass(TypeRef::class.java)
        if (generic !is ParameterizedType) {
            throw IllegalStateException(
                "Reflect to generic superclass of TypeRef failed."
            )
        }
        return generic.actualTypeArguments[0]
    }

    override fun toString(): String {
        return "TypeRef($type)"
    }

    companion object {

        @JvmStatic
        fun <T> of(type: Type): TypeRef<T> {
            return object : TypeRef<T>(type) {}
        }
    }
}

val <T> T.typeRef: TypeRef<T>
    get() {
        return object : TypeRef<T>() {}
    }

fun <T> typeRef(): TypeRef<T> {
    return object : TypeRef<T>() {}
}